# Relational data Exploratory Analysis

In this Exploratory Data Analysis (EDA) project, the primary objective is to hone and demonstrate proficiency in data manipulation techniques while extracting valuable insights from a relational database. By working with structured data spread across multiple related tables, this project provides an opportunity to practice key skills such as data filtering, merging, aggregation, and analysis. The ultimate goal is to transform raw data into meaningful information, uncovering trends, patterns, and relationships that can inform decision-making and provide deeper understanding of the underlying data. 

There are 12 tasks in total and each task aims to be more difficult than the previous one.

In [3]:
# Load libraries
import pandas as pd
import numpy as np
import os

In [6]:
# Load the tables
directory=r'C:\Users\Militsa\Documents\softuni\Projects\dataset'

files=os.listdir(directory)
files

['.ipynb_checkpoints',
 'actor.csv',
 'address.csv',
 'category.csv',
 'city.csv',
 'country.csv',
 'customer.csv',
 'departments.csv',
 'EDA with relational data.ipynb',
 'employee.csv',
 'employees.csv',
 'film.csv',
 'film_actor.csv',
 'film_category.csv',
 'inventory.csv',
 'payment.csv',
 'rental.csv',
 'staff.csv',
 'store.csv']

In [78]:
actor=pd.read_csv('actor.csv')
address=pd.read_csv('address.csv')
category=pd.read_csv('category.csv')
city=pd.read_csv('city.csv')
country=pd.read_csv('country.csv')
customer=pd.read_csv('customer.csv')
departments=pd.read_csv('departments.csv')
film=pd.read_csv('film.csv')
film_actor=pd.read_csv('film_actor.csv')
film_category=pd.read_csv('film_category.csv')
inventory=pd.read_csv('inventory.csv')
payment=pd.read_csv('payment.csv')
rental=pd.read_csv('rental.csv')
staff=pd.read_csv('staff.csv')
store=pd.read_csv('store.csv')


#### Task: Create a list of all the different (distinct) replacement costs of the films. What's the lowest replacement cost?


In [79]:
#The replacement cost values are stored in the film table
sorted_costs=sorted(film['replacement_cost'].unique())
sorted_costs #9.99

[9.99,
 10.99,
 11.99,
 12.99,
 13.99,
 14.99,
 15.99,
 16.99,
 17.99,
 18.99,
 19.99,
 20.99,
 21.99,
 22.99,
 23.99,
 24.99,
 25.99,
 26.99,
 27.99,
 28.99,
 29.99]

#### Task: Write a code that gives an overview of how many films have replacements costs in the following cost ranges: low: 9.99 - 19.99 medium: 20.00 - 24.99 high: 25.00 - 29.99. Question: How many films have a replacement cost in the "low" group?

In [80]:
film['replacement_cost_cat']=film['replacement_cost'].apply(lambda x: 'low' if x<=19.99 else ('medium' if x<=24.99 else 'high'))
film['replacement_cost_cat']

0         low
1         low
2        high
3      medium
4         low
        ...  
995       low
996       low
997       low
998       low
999       low
Name: replacement_cost_cat, Length: 1000, dtype: object

In [81]:
film_cat_count=film.groupby('replacement_cost_cat').size().reset_index(name='CostGroups')
film_cat_count

replacement_cost_cat  CostGroups
0                 high         236
1                  low         514
2               medium         250

#### Task: Create a list of the film titles including their title, length, and category name ordered descendingly by length. Filter the results to only the movies in the category 'Drama' or 'Sports'. Question: In which category is the longest film and how long is it?

In [82]:
film_merged=pd.merge(film, film_category[['category_id','film_id']],on='film_id')
film_merged=pd.merge(film_merged,category[['name','category_id']],on='category_id')

In [83]:
film_merged[(film_merged['name']=='Drama')| (film_merged['name']=='Sports')][['title','length','name']].sort_values('length',ascending=False)

title  length    name
393     SMOOCHY CONTROL     184  Sports
361       RECORDS ZORRO     182  Sports
371      STAR OPERATION     181  Sports
591       JACKET FRISCO     181   Drama
353  MUSSOLINI SPOILERS     180  Sports
..                  ...     ...     ...
330    CRANES RESERVOIR      57  Sports
368         SENSE GREEK      54  Sports
630   NOTTING SPEAKEASY      48   Drama
332     DIVORCE SHINING      47  Sports
627       KWAI HOMEWARD      46   Drama

[136 rows x 3 columns]

#### Task: Create an overview of how many movies (titles) there are in each category (name).Question: Which category (name) is the most common among the films?

In [84]:
film_merged_count=film_merged.groupby('name').size()
film_merged_count.sort_values(ascending=False)

name
Sports         74
Foreign        73
Family         69
Documentary    68
Animation      66
Action         64
New            63
Drama          62
Games          61
Sci-Fi         61
Children       60
Comedy         58
Classics       57
Travel         57
Horror         56
Music          51
dtype: int64

#### Task: Create an overview of the actors' first and last names and in how many movies they appear in. Question: Which actor is part of most movies??

In [85]:
actor.head()

actor_id first_name     last_name             last_update
0         1   PENELOPE       GUINESS  2020-02-15 10:34:33+01
1         2       NICK      WAHLBERG  2020-02-15 10:34:33+01
2         3         ED         CHASE  2020-02-15 10:34:33+01
3         4   JENNIFER         DAVIS  2020-02-15 10:34:33+01
4         5     JOHNNY  LOLLOBRIGIDA  2020-02-15 10:34:33+01

In [86]:
actor_merged=pd.merge(film_actor,actor[['first_name','last_name','actor_id']],on='actor_id')
actor_merged.groupby(['first_name','last_name']).size().sort_values(ascending=False)

first_name  last_name
SUSAN       DAVIS        54
GINA        DEGENERES    42
WALTER      TORN         41
MARY        KEITEL       40
MATTHEW     CARREY       39
                         ..
SISSY       SOBIESKI     18
JULIA       ZELLWEGER    16
            FAWCETT      15
JUDY        DEAN         15
EMILY       DEE          14
Length: 199, dtype: int64

#### Task: Create an overview of the addresses that are not associated to any customer. Question: How many addresses are that?

In [87]:
#Perform a left join between the address DataFrame and the customer DataFrame on the address_id column.
address_customer=pd.merge(address,customer,how='left',on='address_id')
#Filter the results where the first_name column from the customer DataFrame is NaN (which indicates that there is no associated customer for that address).
no_addresses=address_customer[address_customer['first_name'].isna()]
#Count the number of such addresses.
print(no_addresses.shape[0])
print(no_addresses)

4
   address_id               address  address2 district  city_id  postal_code  \
0           1     47 MySakila Drive       NaN  Alberta      300          NaN   
1           2    28 MySQL Boulevard       NaN      QLD      576          NaN   
2           3     23 Workhaven Lane       NaN  Alberta      300          NaN   
3           4  1411 Lillydale Drive       NaN      QLD      576          NaN   

          phone           last_update_x  customer_id  store_id first_name  \
0           NaN  2020-02-15 10:45:30+01          NaN       NaN        NaN   
1           NaN  2020-02-15 10:45:30+01          NaN       NaN        NaN   
2  1.403334e+10  2020-02-15 10:45:30+01          NaN       NaN        NaN   
3  6.172236e+09  2020-02-15 10:45:30+01          NaN       NaN        NaN   

  last_name email activebool create_date last_update_y  active initials  
0       NaN   NaN        NaN         NaN           NaN     NaN      NaN  
1       NaN   NaN        NaN         NaN           NaN     NaN 

#### Task: Create the overview of the sales to determine from which city (we are interested in the city in which the customer lives, not where the store is) most sales occur. Question: What city is that and how much is the amount?


In [88]:
customer_payment_merged=pd.merge(payment, customer,on='customer_id')
customer_payment_merged=pd.merge(customer_payment_merged, address,on='address_id')
customer_payment_merged=pd.merge(customer_payment_merged, city,on='city_id')
customer_payment_merged

payment_id  customer_id  staff_id  rental_id  amount  \
0           16050          269         2          7    1.99   
1           16051          269         1         98    0.99   
2           16052          269         2        678    6.99   
3           16053          269         2        703    0.99   
4           16054          269         1        750    4.99   
...           ...          ...       ...        ...     ...   
16042       31188          195         2       8280    4.99   
16043       31189          195         2       8479    0.99   
16044       31190          195         2       9188    6.99   
16045       31191          195         1       9870    5.99   
16046       31192          195         1       9994    4.99   

                        payment_date  store_id first_name last_name  \
0      2020-01-24 22:40:19.996577+01         1  CASSANDRA   WALTERS   
1      2020-01-25 16:16:50.996577+01         1  CASSANDRA   WALTERS   
2      2020-01-28 22:44:14.996577+01         1  CASSANDRA   WALTERS   
3      2020-01-29 01:58:02.996577+01         1  CASSANDRA   WALTERS   
4      2020-01-29 09:10:06.996577+01         1  CASSANDRA   WALTERS   
...                              ...       ...        ...       ...   
16042  2020-04-29 01:14:17.996577+02         1    VANESSA      SIMS   
16043  2020-04-29 08:10:30.996577+02         1    VANESSA      SIMS   
16044  2020-04-30 11:48:20.996577+02         1    VANESSA      SIMS   
16045  2020-04-30 12:51:17.996577+02         1    VANESSA      SIMS   
16046  2020-04-30 16:58:57.996577+02         1    VANESSA      SIMS   

                                      email  ...  \
0      CASSANDRA.WALTERS@sakilacustomer.org  ...   
1      CASSANDRA.WALTERS@sakilacustomer.org  ...   
2      CASSANDRA.WALTERS@sakilacustomer.org  ...   
3      CASSANDRA.WALTERS@sakilacustomer.org  ...   
4      CASSANDRA.WALTERS@sakilacustomer.org  ...   
...                                     ...  ...   
16042       VANESSA.SIMS@sakilacustomer.org  ...   
16043       VANESSA.SIMS@sakilacustomer.org  ...   
16044       VANESSA.SIMS@sakilacustomer.org  ...   
16045       VANESSA.SIMS@sakilacustomer.org  ...   
16046       VANESSA.SIMS@sakilacustomer.org  ...   

                             address  address2             district city_id  \
0                920 Kumbakonam Loop       NaN           California     446   
1                920 Kumbakonam Loop       NaN           California     446   
2                920 Kumbakonam Loop       NaN           California     446   
3                920 Kumbakonam Loop       NaN           California     446   
4                920 Kumbakonam Loop       NaN           California     446   
...                              ...       ...                  ...     ...   
16042  1792 Valle de la Pascua Place       NaN  Nordrhein-Westfalen     477   
16043  1792 Valle de la Pascua Place       NaN  Nordrhein-Westfalen     477   
16044  1792 Valle de la Pascua Place       NaN  Nordrhein-Westfalen     477   
16045  1792 Valle de la Pascua Place       NaN  Nordrhein-Westfalen     477   
16046  1792 Valle de la Pascua Place       NaN  Nordrhein-Westfalen     477   

       postal_code         phone           last_update_y     city country_id  \
0          75090.0  6.850107e+11  2020-02-15 10:45:30+01  Salinas        103   
1          75090.0  6.850107e+11  2020-02-15 10:45:30+01  Salinas        103   
2          75090.0  6.850107e+11  2020-02-15 10:45:30+01  Salinas        103   
3          75090.0  6.850107e+11  2020-02-15 10:45:30+01  Salinas        103   
4          75090.0  6.850107e+11  2020-02-15 10:45:30+01  Salinas        103   
...            ...           ...                     ...      ...        ...   
16042      15540.0  4.194196e+11  2020-02-15 10:45:30+01   Siegen         38   
16043      15540.0  4.194196e+11  2020-02-15 10:45:30+01   Siegen         38   
16044      15540.0  4.194196e+11  2020-02-15 10:45:30+01   Siegen         38   
16045      15540.0  4.1

In [89]:
sales_overview=customer_payment_merged.groupby('city').sum('amount')
sales_overview['amount'].sort_values(ascending=False)

city
Cape Coral     221.55
Saint-Denis    216.54
Aurora         198.50
Molodetno      195.58
Apeldoorn      194.61
                ...  
Tete            58.82
al-Qadarif      57.81
Bydgoszcz       52.88
Fuzhou          50.86
Tallahassee     50.85
Name: amount, Length: 597, dtype: float64

#### Task: Create an overview of the revenue (sum of amount) grouped by a column in the format "country, city". Question: Which country, city has the least sales?

In [90]:
customer_payment_merged=pd.merge(customer_payment_merged,country[['country','country_id']],on='country_id')


In [94]:
customer_payment_merged['city_country']=customer_payment_merged['city']+" "+customer_payment_merged['country']
customer_payment_merged['city_country']

0        Salinas United States
1        Salinas United States
2        Salinas United States
3        Salinas United States
4        Salinas United States
                 ...          
16042        Kabul Afghanistan
16043        Kabul Afghanistan
16044        Kabul Afghanistan
16045        Kabul Afghanistan
16046        Kabul Afghanistan
Name: city_country, Length: 16047, dtype: object

In [98]:
country_overview=customer_payment_merged.groupby('city_country').sum('amount').sort_values('amount')

In [99]:
country_overview['amount']

city_country
Tallahassee United States      50.85
Fuzhou China                   50.86
Bydgoszcz Poland               52.88
al-Qadarif Sudan               57.81
Tete Mozambique                58.82
                               ...  
Santa Brbara dOeste Brazil    194.61
Molodetno Belarus             195.58
Aurora United States          198.50
Saint-Denis Runion            216.54
Cape Coral United States      221.55
Name: amount, Length: 597, dtype: float64

#### Task: Create a list with the average of the sales amount each staff_id has per customer. Question: Which staff_id makes on average more revenue per customer?

In [108]:
#Group by customer_id and staff_id in the payment DataFrame to calculate the sum of amount for each combination of customer_id and staff_id.
sum_sales_staff=payment.groupby(['staff_id','customer_id'])['amount'].sum().reset_index(name='total')
print(sum_sales_staff)
#Group by staff_id to calculate the average of these sums (i.e., average revenue per customer per staff).
avg_sales_staff=sum_sales_staff.groupby('staff_id')['total'].mean().reset_index(name='avg_amount')
print(avg_sales_staff)

      staff_id  customer_id  total
0            1            1  64.83
1            1            2  60.85
2            1            3  64.86
3            1            4  49.88
4            1            5  73.83
...        ...          ...    ...
1193         2          595  43.87
1194         2          596  41.89
1195         2          597  49.89
1196         2          598  39.88
1197         2          599  54.89

[1198 rows x 3 columns]
   staff_id  avg_amount
0         1   55.907312
1         2   56.632805


#### Task: Create a piece of code that shows average daily revenue of all Sundays. Question: What is the daily average revenue of all Sundays?

In [112]:
# Step 1: Ensure 'payment_date' is a datetime object
payment['payment_date'] = pd.to_datetime(payment['payment_date'],utc=True)

# Step 2: Extract the day of the week (0 = Sunday, 6 = Saturday) and filter for Sundays
payment['weekday'] = payment['payment_date'].dt.dayofweek
sundays = payment[payment['weekday'] == 6]  # Note: Pandas uses 0=Monday, 6=Sunday

# Step 3: Group by date to calculate total revenue for each Sunday
sunday_revenue = sundays.groupby(sundays['payment_date'].dt.date)['amount'].sum().reset_index(name='total')

# Step 4: Calculate the average daily revenue of all Sundays
avg_sunday_revenue = sunday_revenue['total'].mean()

In [113]:
avg_sunday_revenue

1435.15

#### Task: Create a list of movies - with their length and their replacement cost - that are longer than the average length in each replacement cost group. Question: Which two movies are the shortest on that list and how long are they?

In [125]:
#Find the average length in each replacement cost group
cost_groups=film.groupby('replacement_cost')['length'].mean().reset_index(name='avg_length')

#Filter the films with > replacement cost than the average
film_cost_groups=pd.merge(film,cost_groups[['replacement_cost','avg_length']],on='replacement_cost')
films_filtered=film_cost_groups[film_cost_groups['avg_length']<film_cost_groups['length']]

#Sort by length
films_filtered[['title','length','avg_length']].sort_values('length')

title  length  avg_length
409        CELEBRITY HORN     110  109.605263
79    SEATTLE EXPECATIONS     110  107.357143
543  WONDERLAND CHRISTMAS     111  107.440000
53           WIND PHANTOM     111  108.763636
29             SUIT WALLS     111  108.763636
..                    ...     ...         ...
880          POND SEATTLE     185  115.255814
960        CONTROL ANTHEM     185  112.487805
662           GANGS PRIDE     185  114.698113
675     SWEET BROTHERHOOD     185  114.698113
78            SATURN NAME     192  107.357143

[490 rows x 3 columns]

#### Task: Create a list that shows the "average customer lifetime value" grouped by the different districts.
--Example: If there are two customers in "District 1" where one customer has a total (lifetime) spent of 1000 dollars and the second customer has a total spent of $2000 then the "average customer lifetime spent" in this district is 1500 dollars. So, first, you need to calculate the total per customer and then the average of these totals per district.
--Question: Which district has the highest average customer lifetime value?


In [142]:
#Merge payment, customer, address
pay_cus_ad=pd.merge(payment,customer,on='customer_id')
pay_cus_ad=pd.merge(pay_cus_ad,address,on='address_id')

#Calculate the total per customer
total_customer=pay_cus_ad.groupby(['customer_id','district'])['amount'].sum().reset_index(name='total')
total_customer

#Find the average from these totals per district
avg_value_per_district=total_customer.groupby('district')['total'].mean().reset_index(name='avg_customer_spent')
avg_value_per_district.sort_values('avg_customer_spent', ascending=False)

district  avg_customer_spent
278         Saint-Denis              216.54
200               Minsk              195.58
300              Skikda              173.63
159             Khartum              169.65
256             Pietari              162.62
..                  ...                 ...
26          Basel-Stadt               58.86
334                Tete               58.82
372          al-Qadarif               57.81
168  Kujawsko-Pomorskie               52.88
96               Fujian               50.86

[375 rows x 2 columns]